presto-main/src/main/java/com/facebook/presto/server/DynamicFilterService.java

```diff
@@ -0,0 +1,300 @@
+    {
+        dynamicFilterSummaries.entrySet().removeIf(entry -> entry.getKey().getQueryId().equals(queryId));
+        futures.entrySet().removeIf(entry -> entry.getKey().getQueryId().equals(queryId));
+    }
+
+    @Immutable
+    private static class SourceDescriptor
+    {
+        private final String queryId;
+        private final String source;
+
+        public static SourceDescriptor of(String queryId, String source)
+        {
+            return new SourceDescriptor(queryId, source);
+        }
+
+        private SourceDescriptor(String queryId, String source)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.source = requireNonNull(source, "source is null");
+        }
+
+        public String getQueryId()
+        {
+            return queryId;
+        }
+
+        public String getSource()
+        {
+            return source;
+        }
+
+        @Override
+        public boolean equals(Object other)
+        {
+            if (other == this) {
+                return true;
+            }
+            if (other == null || !(other instanceof SourceDescriptor)) {
+                return false;
+            }
+
+            SourceDescriptor sourceDescriptor = (SourceDescriptor) other;
+
+            return Objects.equals(queryId, sourceDescriptor.queryId) &&
+                    Objects.equals(source, sourceDescriptor.source);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(queryId, source);
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .addValue(queryId)
+                    .addValue(source)
+                    .toString();
+        }
+    }
+
+    @NotThreadSafe
+    static class DynamicFilterSummaryWithSenders
+    {
+        private final Map<StageTaskKey, SenderStats> senderStats = new HashMap<>();
+        private final Set<StageTaskKey> registeredTasks;
+        private DynamicFilterSummary dynamicFilterSummary;
+
+        DynamicFilterSummaryWithSenders(Set<StageTaskKey> senderStats)
+        {
+            this.registeredTasks = ImmutableSet.copyOf(senderStats);
+        }
+
+        public Optional<DynamicFilterSummary> getSummaryIfReady()
+        {
+            if (!senderStats.keySet().equals(registeredTasks)) {
+                return Optional.empty();
+            }
+
+            for (Map.Entry<StageTaskKey, SenderStats> entry : senderStats.entrySet()) {
+                if (!entry.getValue().isCompleted()) {
+                    return Optional.empty();
+                }
+            }
+
+            return Optional.of(dynamicFilterSummary);
+        }
+
+        public DynamicFilterSummary addSummary(StageTaskKey stageTaskId, int driverId, DynamicFilterSummary summary, int expectedSummariesCount)
+        {
+            SenderStats stats = senderStats.get(stageTaskId);
```

Comment:
senderStats map is always empty... It is not being filled anywhere in this class. Should we fill it with empty SenderStats at the time of initialisation for all the tasks ?